### PR TITLE
[#277] Add support for adding, removing, and modifying zones (main)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: endpoints/zones/src/main.cpp [hunk 1]
```diff
@@ -9,78 +9,468 @@
...
+							conn, name_iter->second, adm::connection_info_property{value_iter->second});
+					}
+					else if (property_iter->second == "comment") {
+						adm::client::modify_zone(conn, name_iter->second, adm::comment_property{value_iter->second});
+					}
+					else if (property_iter->second == "acl") {
+						// The string referenced by the value iterator is expected to have the following structure:
+					    //
+					    //     <acl>:<user>
+					    //
```
Comment (korydraughn) [inline endpoints/zones/src/main.cpp:256]: Still need to think about the following:
- Should this be exposed at all? _(Maybe. It is a valid operation.)_
- Should consider changing how the input argument is provided (i.e. `<acl>:<user>`)

File: endpoints/zones/src/main.cpp [hunk 2]
```diff
@@ -9,78 +9,468 @@
...
+								acl_user_pair);
+							return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
+						}
+
+						const auto acl_sv = acl_user_pair.substr(0, colon_pos);
+
+						// Extract username from input string.
+					    // TODO Is there a safer way to check this? It's possible the string could exceed the maximum
+					    // allowed size.
+						if (colon_pos + 1 >= acl_user_pair.size()) {
```
Comment (korydraughn) [inline endpoints/zones/src/main.cpp:274]: This isn't so bad given the fact that the HTTP API enforces a max size on the request body.

There's probably a better way to enforce this though.

File: test/test_irods_http_api.py [hunk 3]
```diff
@@ -3339,6 +3339,90 @@ def test_report_operation(self):
...
+            # The properties to update.
+            property_map = [
+                ('name',            'other_zone_renamed'),
+                ('connection_info', 'example.org:1247'),
+                ('comment',         'updated comment'),
+                # The following lines are special. They modify the permissions of
+                # the remote zone for visibility purposes (i.e. strict ACLs). See
+                # the help text of "iadmin modzonecollacl" for more details.
+                #('acl',             'read:rods'),
+                #('acl',             'null:rods')
```
Comment (korydraughn) [inline test/test_irods_http_api.py:3382]: Revisit this.

File: test/test_irods_http_api.py [hunk 4]
```diff
@@ -3339,6 +3339,90 @@ def test_report_operation(self):
...
+                self.assertEqual(r.status_code, 200)
+                self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+                # Capture the new name of the zone following its renaming.
+                if 'name' == p:
+                    zone_name = v
+
+                if 'acl' == p:
+                    # TODO Check the permissions of the "/" collection using the rodsuser.
+                    continue
```
Comment (korydraughn) [inline test/test_irods_http_api.py:3403]: Handle TODO

File: test/test_irods_http_api.py [hunk 5]
```diff
@@ -3339,6 +3339,90 @@ def test_report_operation(self):
...
+            self.assertEqual(result['exists'], True)
+            self.assertEqual(result['info']['name'], zone_name)
+            self.assertEqual(result['info']['type'], 'remote')
+            self.assertEqual(result['info']['connection_info'], '')
+            self.assertEqual(result['info']['comment'], '')
+
+            # The properties to update.
+            property_map = [
+                ('name',            'other_zone_renamed'),
+                ('connection_info', 'example.org:1247'),
```
Comment (alanking) [inline test/test_irods_http_api.py:3376]: Consider adding a test which feeds bad inputs (missing colon, empty string on either side of the colon, etc.).

File: test/test_irods_http_api.py [hunk 6]
```diff
@@ -3339,6 +3339,90 @@ def test_report_operation(self):
         self.assertIn('zones', zone_report)
         self.assertGreaterEqual(len(zone_report['zones']), 1)
 
+    def test_adding_removing_and_modifying_zones(self):
+        headers = {'Authorization': 'Bearer ' + self.rodsadmin_bearer_token}
```
Comment (alanking) [inline test/test_irods_http_api.py:3343]: Is there any value in showing that non-admin users cannot perform any of these action (as is being done in the test for adding, modifying, and removing resources)? "No" is an okay answer here since I'm guessing (hoping) the administration library or the admin API is enforcing this.

**Review by alanking — COMMENTED** [review]
Looking good so far

File: endpoints/zones/src/main.cpp [hunk 7]
```diff
@@ -9,78 +9,468 @@
...
+
+using json = nlohmann::json;
+// clang-format on
+
+#define IRODS_HTTP_API_ENDPOINT_OPERATION_SIGNATURE(name) \
+  auto name(                                              \
+	  irods::http::session_pointer_type _sess_ptr,        \
+	  irods::http::request_type& _req,                    \
+	  irods::http::query_arguments_type& _args)           \
+	  ->void
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply endpoints/zones/src/main.cpp:39]: That is how clang-format chose to format it. It matches other locations in the codebase.

I'm trying not to get distracted by things like that so leaving as is.

File: endpoints/zones/src/main.cpp [hunk 8]
```diff
@@ -9,78 +9,468 @@
...
+				try {
+					const auto name_iter = _args.find("name");
+					if (name_iter == std::end(_args)) {
+						logging::error("{}: Missing [name] parameter.", fn);
+						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
+					}
+
+					adm::zone_options opts;
+
+					auto connection_info_iter = _args.find("connection-info");
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply endpoints/zones/src/main.cpp:114]: Good catch. They shouldn't be const since I call `std::move` on `->second`.

[the remark this answers is not in the document]
Comment (korydraughn) [reply endpoints/zones/src/main.cpp:114]: Done.

File: endpoints/zones/src/main.cpp [hunk 9]
```diff
@@ -9,78 +9,468 @@
...
+			return _sess_ptr->send(std::move(*result.response));
+		}
+
+		const auto client_info = result.client_info;
+
+		irods::http::globals::background_task([fn = __func__,
+		                                       client_info,
+		                                       _sess_ptr,
+		                                       _req = std::move(_req),
+		                                       _args = std::move(_args)] {
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply endpoints/zones/src/main.cpp:408]: Most of the code is formatted in that way thanks to clang-format.

Not my favorite, but also not incorrect.

File: test/test_irods_http_api.py (hunk 6 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3343]: Yes, the test should do that regardless of the internals just so we don't go backwards. We know non-admin users cannot do it today due to the current iRODS behavior, but that may not be the case a year (or more) from now.

File: test/test_irods_http_api.py (hunk 5 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3376]: Done

File: endpoints/zones/src/main.cpp (hunk 1 above)
Comment (korydraughn) [reply endpoints/zones/src/main.cpp:255]: After some experimentation with `iadmin modzonecollacl`, adding a dedicated operation for it makes the most sense.

Removing.

File: test/test_irods_http_api.py (hunk 3 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3382]: After some experimentation with `iadmin modzonecollacl`, adding a dedicated operation for it makes the most sense.

Removing.

File: test/test_irods_http_api.py (hunk 4 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3403]: See https://github.com/irods/irods_client_http_api/pull/282#discussion_r1667147855

File: endpoints/zones/src/main.cpp (hunk 1 above)
Comment (korydraughn) [reply endpoints/zones/src/main.cpp]: Done.

File: test/test_irods_http_api.py (hunk 3 above)
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done.

File: test/test_irods_http_api.py (hunk 4 above)
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done.

File: test/test_irods_http_api.py (hunk 6 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3347]: Done

File: test/test_irods_http_api.py [hunk 10]
```diff
@@ -3339,6 +3343,246 @@ def test_report_operation(self):
...
+        # Add a remote zone to the local zone.
+        zone_name = 'tempZone'
+        #zone_name = 'other_zone'
+        #r = requests.post(self.url_endpoint, headers=rodsadmin_headers, data={
+        #    'op': 'add',
+        #    'name': zone_name
+        #})
+        #self.logger.debug(r.content)
+        #self.assertEqual(r.status_code, 200)
+        #self.assertEqual(r.json()['irods_response']['status_code'], 0)
```
Comment (korydraughn) [inline test/test_irods_http_api.py:3505]: Does `modzonecollacl` require a remote zone? The docs kind of imply it does, but testing with `iadmin` and using the zone administration lib seem to not enforce that.

Need to test the iRODS API more.

File: test/test_irods_http_api.py [hunk 11]
```diff
@@ -3339,6 +3343,246 @@ def test_report_operation(self):
...
+                'op': 'set_zone_collection_permission',
+                'name': zone_name,
+                'user': self.rodsuser_username,
+                'permission': 'null'
+            })
+            # Remove the remote zone.
+            #r = requests.post(self.url_endpoint, headers=rodsadmin_headers, data={
+            #    'op': 'remove',
+            #    'name': zone_name
+            #})
```
Comment (korydraughn) [inline test/test_irods_http_api.py:3551]: Don't forget to handle this.

File: endpoints/zones/src/main.cpp (hunk 2 above)
Comment (korydraughn) [reply endpoints/zones/src/main.cpp]: Removed.

File: endpoints/zones/src/main.cpp [hunk 12]
```diff
@@ -248,58 +250,94 @@ namespace
...
+					}
+
+#if 0
+					// TODO Are zones supported?
+					const auto user_zone_iter = _args.find("user_zone");
+					if (value_iter == std::end(_args)) {
+						logging::error("{}: Missing [user] parameter.", fn);
 						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
 					}
+#endif
```
Comment (alanking) [inline endpoints/zones/src/main.cpp]: Guessing this is an ongoing investigation?

File: test/test_irods_http_api.py [hunk 13]
```diff
@@ -3493,13 +3489,95 @@ def test_setting_invalid_connection_info_results_in_an_error(self):
...
+        #})
+        #self.logger.debug(r.content)
+        #self.assertEqual(r.status_code, 200)
+        #self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+        try:
+            # Show the rodsuser cannot list the contents of the zone collection.
+            r = requests.get(f'{self.url_base}/collections', headers=rodsuser_headers, params={
+                'op': 'list',
+                #'lpath': '/'
```
Comment (alanking) [inline test/test_irods_http_api.py:3511]: Left here intentionally?

File: endpoints/zones/src/main.cpp (hunk 12 above)
Comment (korydraughn) [reply endpoints/zones/src/main.cpp]: Correct.

File: test/test_irods_http_api.py (hunk 13 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3511]: Yes. A leftover due to investigation.

File: test/test_irods_http_api.py (hunk 10 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3505]: The local zone is adequate for this test. Remote zones aren't required to prove correctness.

This was determined through local bench testing.

Comment (korydraughn) [reply test/test_irods_http_api.py:3505]: On second thought, more investigation of `iadmin modzonecollacl` is needed before making this operation available.

Disabling the operation and tests until further notice.

I'll open an issue to revisit it later. No need to hold up the PR for this one operation.

File: test/test_irods_http_api.py (hunk 11 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:3551]: See https://github.com/irods/irods_client_http_api/pull/282#discussion_r1669318629.

File: endpoints/zones/src/main.cpp (hunk 12 above)
Comment (korydraughn) [reply endpoints/zones/src/main.cpp]: See https://github.com/irods/irods_client_http_api/pull/282#discussion_r1669318629.

File: test/test_irods_http_api.py [hunk 14]
```diff
@@ -3456,7 +3451,8 @@ def test_setting_invalid_connection_info_results_in_an_error(self):
             value_error_map = [
                 ('example.org:', irods_error_codes.CAT_INVALID_ARGUMENT),
                 (':1247', irods_error_codes.CAT_HOSTNAME_INVALID),
-                (':', irods_error_codes.CAT_HOSTNAME_INVALID)
+                (':', irods_error_codes.CAT_HOSTNAME_INVALID),
+                ('', irods_error_codes.CAT_INVALID_ARGUMENT)
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done

File: endpoints/zones/src/main.cpp [hunk 15]
```diff
@@ -310,21 +314,13 @@ namespace
 						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
 					}
 
+					// TODO(#???) Investigate support for fully-qualified usernames do not appear to be supported.
```
Comment (alanking) [inline endpoints/zones/src/main.cpp:317]: Guessing we'll need an issue number here.

**Review by alanking — COMMENTED** [review]
Nice. Just saw one more thing to tweak (maybe) and I think this is good to go.

File: endpoints/zones/src/main.cpp (hunk 15 above)
Comment (alanking) [reply endpoints/zones/src/main.cpp]: Looks like you got this covered. Resolving.

**Review by alanking — COMMENTED** [review]
Is there anything else to do here? Tests are passing? If so, please squash to taste and we'll do a once-over before #s.

**Comment by korydraughn** [comment]
Tests pass. This is ready.

**Comment by korydraughn** [comment]
#'d

**Comment by korydraughn** [comment]
Merging.
